aea/test_tools/test_skill.py

```diff
@@ -0,0 +1,402 @@
+            else dialogue.last_message.message_id + 1
+        )
+        target = target if target is not None else dialogue.last_message.message_id
+        to = to if to is not None else dialogue.self_address
+        sender = (
+            sender
+            if sender is not None
+            else dialogue.dialogue_label.dialogue_opponent_addr
+        )
+
+        incoming_message = self.build_incoming_message(
+            message_type=message_type,
+            performative=performative,
+            dialogue_reference=dialogue_reference,
+            message_id=message_id,
+            target=target,
+            to=to,
+            sender=sender,
+            **kwargs,
+        )
+
+        return incoming_message
+
+    @staticmethod
+    def _provide_unspecified_fields(
+        message: DialogueMessage, last_is_incoming: Optional[bool], message_id: int
+    ) -> Tuple[bool, int]:
+        """
+        Specifies values (an interpretation) for the unspecified fields of a DialogueMessage.
+
+        For an unspecified is_incoming, the opposite of the last_is_incoming value is used.
+        For an unspecified target, the message_id of the previous message (message_id - 1) is used.
+
+        :param message: the DialogueMessage
+        :param last_is_incoming: the is_incoming value of the previous DialogueMessage
+        :param message_id: the message_id of this DialogueMessage
+
+        :return: the is_incoming and target values
+        """
+        default_is_incoming = not last_is_incoming
+        is_incoming = default_is_incoming if message[2] is None else message[2]
+
+        default_target = message_id - 1
+        target = default_target if message[3] is None else message[3]
+        return is_incoming, target
+
+    @staticmethod
+    def _non_initial_incoming_message_dialogue_reference(
+        dialogue: Dialogue,
+    ) -> Tuple[str, str]:
+        """
+        Specifies the dialogue reference of a non-initial incoming message for a dialogue.
+
+        It uses a complete version of the reference in the dialogue if it is incomplete,
+        otherwise it uses the reference in the dialogue.
+
+        :param dialogue: the dialogue to which the incoming message is intended
+        :return: its dialogue reference
+        """
+        dialogue_reference = (
+            dialogue.dialogue_label.dialogue_reference[0],
+            Dialogues._generate_dialogue_nonce()  # pylint: disable=protected-access
+            if dialogue.dialogue_label.dialogue_reference[1]
+            == Dialogue.UNASSIGNED_DIALOGUE_REFERENCE
+            else dialogue.dialogue_label.dialogue_reference[1],
+        )
+        return dialogue_reference
+
+    def _extract_message_fields(
+        self,
+        message: DialogueMessage,
+        messages: Tuple[DialogueMessage, ...],
+        last_is_incoming: bool,
+    ):
+        performative = message[0]
+        contents = message[1]
+        message_id = messages.index(message) + 1
```

Comment:
erm just provide the index in the args? Decouples the logic a bit.
```suggestion
        message_id = index + 1
```

Reply:
Agree. Done.